src/test/common/test_random.cc

```diff
@@ -0,0 +1,274 @@
+  }
+ 
+  // Multiple types (floating point):
+  {
+  double min = 0.0, max = 1.0;
+  type_check_ok(min, max);
+  }
+ 
+  {
+  float min = 0.0, max = 1.0;
+  type_check_ok(min, max);
+  }
+ 
+  // min > max should not explode:
+  {
+  float min = 1.0, max = 0.0;
+  auto x = ceph::util::generate_random_number(min, max);
+  }
+}
+
+TEST(util, test_user_mutex)
+{
+  // Show that we can call with a user lock:
+  std::mutex l;
+ 
+  ceph::util::randomize_rng(l);
+ 
+  ceph::util::generate_random_number(l);
+ 
+  ceph::util::generate_random_number(20, l);
+  ceph::util::generate_random_number(1, 20, l);
+  ceph::util::generate_random_number(20l, l);
+  ceph::util::generate_random_number(1l, 20l, l);
+ 
+  ceph::util::generate_random_number(20.0, l);
+  ceph::util::generate_random_number(1.0, 20.0, l);
+  ceph::util::generate_random_number(double(20.0), l);
+  ceph::util::generate_random_number(double(1.0), double(20.0), l);
+}
+
+TEST(util, test_random_user_rng_lock)
+{
+  /* Users may not want to use the static RNG-- for example,
+  in a case where the developer would like one RNG per thread. For
+  now, we support a custom RNG only along with an accompanying
+  custom lock through the function interface: */
+ 
+  std::mutex l;
+  std::mt19937_64 e;
+ 
+  ceph::util::randomize_rng(l, e);
+ 
+  ceph::util::generate_random_number(l, e);
+ 
+  ceph::util::generate_random_number(20, l, e);
+  ceph::util::generate_random_number(1, 20, l, e);
+ 
+  ceph::util::generate_random_number(20l, l, e);
+ 
+  ceph::util::generate_random_number(1l, 20l, l, e);
+ 
+  ceph::util::generate_random_number(20.0, l, e);
+  ceph::util::generate_random_number(1.0, 20.0, l, e);
+ 
+  ceph::util::generate_random_number(double(20.0), l, e);
+  ceph::util::generate_random_number(double(1.0), double(20.0), l, e);
+}
+
+TEST(util, test_make_random_function)
+{
+  // Test convenience functions:
+ 
+  constexpr auto max_int   = std::numeric_limits<int>::max();
+ 
+  {
+  auto rng_i = ceph::util::make_random_number_function<1, max_int - 1>();
+ 
+  int x = rng_i(), y = rng_i();
+ 
+  ASSERT_GT(x, 0);
+  ASSERT_LT(x, max_int);
+  ASSERT_GT(y, 0);
+  ASSERT_LT(y, max_int);
+  ASSERT_NE(x, y);
+  }
+  
+  // User may specify engine:
+  {
+  auto rng_i = ceph::util::make_random_number_function<1, 10, std::mt19937_64>();
+  int x = rng_i();
+  ASSERT_GT(x, 0); 
+  ASSERT_LT(x, 10); 
+  }
+ 
+  // User specified seed:
+  {
+  auto rng_i = ceph::util::make_random_number_function<1, 10, 42>();
+  int x = rng_i();
+  ASSERT_GT(x, 0); 
+  ASSERT_LT(x, 10); 
+  }
+}
+
+TEST(util, test_random_class_interface)
+{
+  ceph::util::random_number_generator<int> rng;
+ 
+  // Other ctors:
+  {
+  ceph::util::random_number_generator<int> rng(1234);   // seed
+  }
+ 
+  {
+  auto a = rng();
+  auto b = rng();
+ 
+  // Technically can fail, but should "almost never" happen:
+  ASSERT_NE(a, b);
+  }
+ 
+  {
+  auto a = rng(10);
+  ASSERT_LT(a, 11);
+  ASSERT_GT(a, -1);
+  }
+ 
+  {
+  auto a = rng(10.0);
```

Comment:
In this case, I'd expect variable 'a' to be a float. It's just checking that the mechanism compiles. The value should be between 0.0 and 10.0.